resources/net.go

```diff
@@ -0,0 +1,849 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/purpleidea/mgmt/recwatch"
+
+	multierr "github.com/hashicorp/go-multierror"
+	errwrap "github.com/pkg/errors"
+	// XXX: Do NOT use subscribe methods from this lib, as they are racey and
+	// do not clean up spawned goroutines. Should be replaced when a suitable
+	// alternative is available.
+	"github.com/vishvananda/netlink"
+)
+
+func init() {
+	RegisterResource("net", func() Res { return &NetRes{} })
+}
+
+const (
+	// ifaceUp is the up (on) interface state.
+	ifaceUp = "up"
+	// ifaceDown is the down (off) interface state.
+	ifaceDown = "down"
+
+	// netlink multicast groups to watch for events
+	rtmGrpLink        = 0x1   // interface create/delete/up/down
+	rtmGrpIPv4IfAddr  = 0x10  // add/delete IPv4 addresses
+	rtmGrpIPv6IfAddr  = 0x100 // add/delete IPv6 addresses
+	rtmGrpIPv4IfRoute = 0x40  // add delete routes
+
+	// ip routing protocols for used for netlink route messages
+	rtProtoKernel = 2 // kernel
+	rtProtoStatic = 4 // static
+
+	// networkdUnitFileDir is the location of networkd unit files which define
+	// the systemd network connections.
+	networkdUnitFileDir = "/etc/systemd/network/"
+	// IfacePrefix is the prefix used to identify unit files for managed links.
+	IfacePrefix = "mgmt-"
+	// networkdUnitFileExt is the file extension for networkd unit files.
+	networkdUnitFileExt = ".network"
+)
+
+// NetRes is an systemd-networkd resource.
+type NetRes struct {
+	BaseRes `yaml:",inline"`
+	State   string   `yaml:"state"`   // up, down, or empty
+	Addrs   []string `yaml:"addrs"`   // list of addresses in cidr format
+	Gateway string   `yaml:"gateway"` // gateway address
+
+	iface        *iface // a struct containing the net.Interface and netlink.Link
+	unitFilePath string // the interface unit file path
+	// XXX: replace TempDir with VarDir
+	tempDir string // temporary directory for storing the pipe socket file
+}
+
+// nlChanStruct defines the channel used to send netlink messages and errors
+// to the event processing loop in Watch.
+type nlChanStruct struct {
+	msg []syscall.NetlinkMessage
+	err error
+}
+
+// iface wraps net.Interface to add additional methods.
+type iface struct {
+	*net.Interface
```

Comment:
What is the reason this is anonymous?
It's okay if there's a useful reason, if not, it's good to be explicit so we know where a method comes from.
For discussion instead of immediate change.

Reply:
TBH, the struct didn't have the netlink.Link to begin with, and when I added it, I never  gave the *net.Interface a name. I don't have a problem changing it.